### PR TITLE
add conditional logic for 204 status

### PR DESCRIPTION
**Pull request by ngarland9**

### CODE REVIEW

File: lib/rails_response_dumper/runner.rb [hunk 1]
```diff
@@ -60,5 +48,24 @@ def rollback_after
...
       end
     end
+
+    def handle_repsonses(responses, dump_block, dumper_dir)
+      responses.each_with_index do |response, index|
+        if response.status == dump_block.expected_status_code
+          mime = response.content_type.split(/ *; */).first
+          extension = MIME::Types[mime].first.preferred_extension
+          File.write("#{dumper_dir}/#{index}.#{extension}", response.body)
+        elsif response.status == 204
```
Comment (jdufresne) [inline lib/rails_response_dumper/runner.rb:58]: I don't think we should treat 204 as a special case. Instead, we should allow for different responses within the same dump block to have different status codes. Whether they are 204, 302, or 404, or whatever.

Doing so would solve the general problem, not just one surfaced issue.

File: lib/rails_response_dumper/runner.rb [hunk 2]
```diff
@@ -60,5 +48,24 @@ def rollback_after
         yield
       end
     end
+
+    def handle_repsonses(responses, dump_block, dumper_dir)
+      responses.each_with_index do |response, index|
+        if response.status == dump_block.expected_status_code
+          mime = response.content_type.split(/ *; */).first
+          extension = MIME::Types[mime].first.preferred_extension
+          File.write("#{dumper_dir}/#{index}.#{extension}", response.body)
```
Comment (jdufresne) [inline lib/rails_response_dumper/runner.rb:57]: The `File.write` can be factored out of the `if`/`else` branch.

**Review by jdufresne — COMMENTED** [review]
Could you add a test, please?

File: lib/rails_response_dumper/runner.rb [hunk 3]
```diff
@@ -31,16 +31,18 @@ def run_dumps
           FileUtils.mkdir_p dumper_dir
 
           dumper.responses.each_with_index do |response, index|
-            unless response.status == dump_block.expected_status_code
+            if response.content_type.nil?
```
Comment (jdufresne) [inline lib/rails_response_dumper/runner.rb:34]: Even if content type doesn't exist, the status code should be verified.

File: spec/app/dumpers/test_response_dumper.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+ResponseDumper.define 'Test' do
+  dump 'index_mulitple_requests', status_codes: %i[ok no_content] do
+    get root_path
+    put url_for(action: :destroy, controller: 'test')
```
Comment (jdufresne) [inline spec/app/dumpers/test_response_dumper.rb:6]: Could use the path method here.

**Review by jdufresne — COMMENTED** [review]
Thanks! This looks really good overall.

The README should be updated to use the new `status_codes` argument. As well, I think we should add a new example to `HTTP Status Codes` that explains how to use multiple different status codes.

**Comment by jdufresne** [comment]
When this is ready for the next round of review, can you please squash commits and update the commit message to include all the new details with background, description, rationale, etc.?

File: README.md [hunk 5]
```diff
@@ -46,14 +46,26 @@ run in a transaction that always rollsback at the end.
...
 # dumpers/users_response_dumper.rb
 
 ResponseDumper.define 'Users' do
-  dump 'show_does_not_exist', status_code: :not_found do
+  dump 'show_does_not_exist', status_codes: [:not_found] do
+    get user_path(0)
+  end
+end
+```
+If your dumper makes multiple requests you will have to specify each expected response
```
Comment (jdufresne) [inline README.md:61]: Blank line after the code block. please.

File: spec/app/dumpers/test_response_dumper.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+ResponseDumper.define 'Test' do
+  dump 'index_mulitple_requests', status_codes: %i[ok no_content] do
```
Comment (jdufresne) [inline spec/app/dumpers/test_response_dumper.rb:4]: Why is this call "index" if it dumps `destroy`?

```ruby
  dump 'mulitple_requests', status_codes: %i[ok no_content] do
```

File: spec/app/dumpers/test_response_dumper.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+    get root_path
+    delete url_for(action: :destroy, controller: 'test')
```
Comment (jdufresne) [inline spec/app/dumpers/test_response_dumper.rb:6]: IMO, `controller` should come first. Can make a symbol.

```suggestion
    delete url_for(controller: :test, action: :destroy)
```

File: spec/app/config/routes.rb [hunk 8]
```diff
@@ -2,4 +2,6 @@
 
 Rails.application.routes.draw do
   root 'root#index'
+
+  resource :test, controller: :test, only: [:destroy]
```
Comment (jdufresne) [inline spec/app/config/routes.rb:6]: You shouldn't need to specify `controller` if it matches the name of the resource. For details, see: https://guides.rubyonrails.org/routing.html

**Comment by jdufresne** [comment]
@ngarland9 I made some suggested edits in https://github.com/ngarland9/rails-response-dumper/pull/1

If they look good to you, rest looks good to me. Let me know if you'd like to discuss any of them.

**Comment by ngarland9** [comment]
> @ngarland9 I made some suggested edits in [ngarland9#1](https://github.com/ngarland9/rails-response-dumper/pull/1)
> 
> If they look good to you, rest looks good to me. Let me know if you'd like to discuss any of them.

Looks great! Thank you so much! Great catch on the controller name now I see why my route wasn't working! lol

**Comment by jdufresne** [comment]
Code looks good. I think it would be nice to provide additional detail in commit messages. This one is a bit short and doesn't provide any brackground or reason for the change. If this kind of detail could be added regularly, that would be helpful to the review process, but also when looking back at historical changes. Here is a suggested commit message. Feel for to reuse or not or to reword to your liking:

---

Allow multiple expected status codes

To support dump methods with multiple responses that may have different
HTTP status codes, the dump argument "status_code" is now
"status_codes". It accepts an Array instead of a scalar.

To support the HTTP 204 status code (No Content), the Content-Type
header is only parsed if it exists. If it does not exist, the dumped
response does not have a file extension.

**Review by jdufresne — APPROVED** [review]

**Pull request merged**